tests/sentry/web/frontend/test_error_page_embed.py

```diff
@@ -90,6 +93,53 @@ def test_uses_locale_from_header(self):
         self.assertTemplateUsed(resp, "sentry/error-page-embed.html")
         assert "Fermer" in resp.content  # Close
 
+    def test_xss(self):
+        user_feedback_options = {}
+
+        # options that are known to be injected into JS
+        unsafe_option_keys = ["errorFormEntry", "successMessage", "errorGeneric"]
+        for key in unsafe_option_keys:
+            user_feedback_options[key] = "<img src=x onerror=alert({0})>XSS_{0}".format(key)
+
+        # options that are known to be used within a Django template but are automatically escaped
+        safe_option_keys = [
+            "title",
+            "subtitle",
+            "subtitle2",
+            "labelName",
+            "labelEmail",
+            "labelComments",
+            "labelSubmit",
+            "labelClose",
+        ]
+        for key in safe_option_keys:
+            user_feedback_options[key] = mark_safe(
+                "<img src=x onerror=alert({0})>XSS_{0}".format(key)
+            )
+
+        user_feedback_options_qs = urlencode(user_feedback_options)
+        path_with_qs = "%s?eventId=%s&dsn=%s&%s" % (
+            self.path,
+            quote(self.event_id),
+            quote(self.key.dsn_public),
+            user_feedback_options_qs,
+        )
+        resp = self.client.get(
+            path_with_qs,
+            HTTP_REFERER="http://example.com",
+            HTTP_ACCEPT="text/html, text/javascript",
+        )
+        assert resp.status_code == 200, resp.content
+        self.assertTemplateUsed(resp, "sentry/error-page-embed.html")
+
+        for xss_payload in user_feedback_options.values():
+            is_safe = isinstance(xss_payload, SafeData)
+            if is_safe:
+                assert escape(xss_payload) in resp.content, resp.content
+            else:
+                assert json.dumps_htmlsafe(escape(xss_payload)) in resp.content, resp.content
```

Comment:
This is a non-trivial amount of logic in the test assertions. Would it be as effective to ensure that there are no instances of your injected content?

Reply:
@markstory 

That's what I thought initially. 

But I don't think it's sufficient to merely check the absence of the XSS payload string because `json.dumps_htmlsafe` converts `<` and `>` to their unicode escape sequence representations:

```
>>> json.dumps_htmlsafe("<img src=x onerror=alert(2)>XSS_2")
'"\\u003cimg src=x onerror=alert(2)\\u003eXSS_2"'
```

------

Within `error-page-embed.js` (https://github.com/getsentry/sentry/blob/7ee75b9e003bdd688ed0929a74ca9b59e83af870/src/sentry/templates/sentry/error-page-embed.js), it gets injected (without escaping) as follows:

```
  var strings = {"sent_message":"\u003cimg src=x onerror=alert(2)\u003eXSS_2","form_error":"\u003cimg src=x onerror=alert(1)\u003eXSS_1","generic_error":"\u003cimg src=x onerror=alert(3)\u003eXSS_3"};
```

which is still susceptible to XSS. 

To demonstrate this, I've distilled this particular XSS vector as a demo here: https://codepen.io/Dashed/pen/QWwrEbX?editors=1010

Reply:
Would ensuring that both the original string and the `dumps_htmlsafe` version are not present in the output work?

Smashing strings together in the `error-page-embed.js` is the root vector here. Perhaps we should be fixing that instead?

Code like

```javascript
    this._formContent.innerHTML = '<p class="message-success">' + strings.sent_message + '</p>';
```

Could become

```javascript
var message = document.createNode('p');
p.setAttribute('class', 'message-success');
p.appendChild(document.createTextNode(strings.sent_message));
this._formContent.innerHtml = '';
this._formContent.appendChild(message)
```

This would be a bit more verbose, but would be safer long term as there is no chance of an escaping/encoding mistake in the future.

Reply:
That approach is more favourable.